### PR TITLE
Clean up `compiler/rustc*/Cargo.toml`

### PR DESCRIPTION
**Pull request by nnethercote**
Mostly by sorting dependencies, plus some other minor things.

r? @wesleywiser

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

**Comment by klensy** [comment]
https://github.com/rust-lang/rustfmt/issues/4091

**Comment by nnethercote** [comment]
> [rust-lang/rustfmt#4091](https://github.com/rust-lang/rustfmt/issues/4091)

I did a trial run of this PR's approach in #116477 where I said the following:

> There are ongoing efforts to provide linting/formatting of Cargo.toml files, e.g. https://github.com/rust-lang/rustfmt/pull/5240, https://crates.io/crates/cargo-toml-lint, and https://github.com/TimonPost/cargo-toml-format. But it's far from clear what's the right approach.
>
> So this PR does something very simple: it uses the order checking already present in tidy. This allows incremental application of ordering, starting right now, and avoiding the need for any kind of all-at-once conversion.
>
> If we do end up using some more comprehensive Cargo.toml linting/formatting solution in the future, the tidy-alphabetical lines will be easy to remove.

I think that still applies.

File: src/tools/tidy/src/alphabetical.rs [hunk 1]
```diff
@@ -71,7 +76,7 @@ fn check_section<'a>(
         let trimmed_line = line.trim_start_matches(' ');
 
         if trimmed_line.starts_with("//")
-            || trimmed_line.starts_with("#[")
+            || trimmed_line.starts_with("#")
```
Comment (cjgillot) [inline src/tools/tidy/src/alphabetical.rs:79]: We use this to check that feature attributes are sorted. Will this still work?

Comment (nnethercote) [reply src/tools/tidy/src/alphabetical.rs:79]: Good point. As written, no, but it should be possible to tweak the string matching so it does continue to work.

Comment (nnethercote) [reply src/tools/tidy/src/alphabetical.rs:79]: Wait a minute, feature attributes aren't sorted. E.g. `compiler/rustc_arena/lib.rs` is an example without sorting. And this tidy sorting is only between `tidy-alphabetical-{start,end}` markers.

Comment (cjgillot) [reply src/tools/tidy/src/alphabetical.rs:79]: This case for instance: https://github.com/rust-lang/rust/blob/master/library/core/src/lib.rs#L113-L195

Comment (Nilstrieb) [reply src/tools/tidy/src/alphabetical.rs:79]: This is supposed to make it ignore attributes in the features macro. Extending it to cover `#` (toml comments) sounds fine, but should have a comment that it also ignored attributes.

Comment (nnethercote) [reply src/tools/tidy/src/alphabetical.rs]: I have updated the code to ignore lines starting with `#` except for lines starting with `#!`.

**Comment by nnethercote** [comment]
> ##[error]tidy error: /checkout/compiler/rustc_mir_transform/Cargo.toml:11: line not in alphabetical order

I don't understand this error. The line in question looks fine. Also, there was no failure on the initial push, and that file didn't change on the subsequent push.

File: src/tools/tidy/src/alphabetical.rs [hunk 2]
```diff
@@ -100,11 +105,8 @@ pub fn check(path: &Path, bad: &mut bool) {
...
         let mut lines = contents.lines().enumerate().peekable();
         while let Some((_, line)) = lines.next() {
-            if line.contains(START_COMMENT) {
+            if line.contains(START_MARKER) {
                 check_section(file, &mut lines, bad);
-                if lines.peek().is_none() {
-                    tidy_error!(bad, "{file}: reached end of file expecting `{END_COMMENT}`")
-                }
             }
         }
```
Comment (klensy) [inline src/tools/tidy/src/alphabetical.rs:111]: Unused peekable?

Comment (nnethercote) [reply src/tools/tidy/src/alphabetical.rs]: True. I have updated accordingly.

**Comment by nnethercote** [comment]
> > ##[error]tidy error: /checkout/compiler/rustc_mir_transform/Cargo.toml:11: line not in alphabetical order
> 
> I don't understand this error. The line in question looks fine. Also, there was no failure on the initial push, and that file didn't change on the subsequent push.

I added some debugging code that prints out the contents of the file triggering the error:
```
<<<BAD: /checkout/compiler/rustc_mir_transform/Cargo.toml
[package]
##[error]tidy error: /checkout/compiler/rustc_mir_transform/Cargo.toml:11: line not in alphabetical order
name = "rustc_mir_transform"
version = "0.0.0"
edition = "2021"

[dependencies]
# tidy-alphabetical-start
either = "1"
itertools = "0.10.1"
rustc_ast = { path = "../rustc_ast" }
rustc_arena = { path = "../rustc_arena" }
rustc_attr = { path = "../rustc_attr" }
rustc_const_eval = { path = "../rustc_const_eval" }
rustc_data_structures = { path = "../rustc_data_structures" }
rustc_errors = { path = "../rustc_errors" }
rustc_fluent_macro = { path = "../rustc_fluent_macro" }
rustc_hir = { path = "../rustc_hir" }
rustc_index = { path = "../rustc_index" }
rustc_macros = { path = "../rustc_macros" }
rustc_middle = { path = "../rustc_middle" }
rustc_mir_build = { path = "../rustc_mir_build" }
rustc_mir_dataflow = { path = "../rustc_mir_dataflow" }
rustc_serialize = { path = "../rustc_serialize" }
rustc_session = { path = "../rustc_session" }
rustc_span = { path = "../rustc_span" }
rustc_target = { path = "../rustc_target" }
rustc_trait_selection = { path = "../rustc_trait_selection" }
smallvec = { version = "1.8.1", features = ["union", "may_dangle"] }
tracing = "0.1"
# tidy-alphabetical-end

[dev-dependencies]
# tidy-alphabetical-start
coverage_test_macros = { path = "src/coverage/test_macros" }
# tidy-alphabetical-end
>>>BAD
```
It's complaining about the `rustc_arena` occurring after the `rustc_attr` line, which would be fined except that `compiler/rustc_mir_transform/Cargo.toml` **doesn't contain a `rustc_arena` line**! Not in the current repository, and not in the code that forms this PR. Other than that line, the contents of this `Cargo.toml` file looks as expected.

I have absolutely no idea how this line is getting inserted. WTF.

**Comment by cuviper** [comment]
> It's complaining about the `rustc_arena` occurring after the `rustc_ast` line, which would be fine except that `compiler/rustc_mir_transform/Cargo.toml` **doesn't contain a `rustc_arena` line**! Not in the current repository, and not in the code that forms this PR.

#107009 merged with that yesterday:
https://github.com/rust-lang/rust/blob/2e4e2a8f288f642cafcc41fff211955ceddc453d/compiler/rustc_mir_transform/Cargo.toml#L14

**Comment by nnethercote** [comment]
I fixed the conflicts.

File: src/tools/tidy/src/alphabetical.rs (hunk 1 above)
Comment (nnethercote) [reply src/tools/tidy/src/alphabetical.rs]: Note: thanks to #117149 landing, we now have unit tests that would have caught this problem.

**Review by wesleywiser — APPROVED** [review]

**Pull request merged**